toolchain/check/testdata/impl/no_prelude/fail_impl_as_scope.carbon

```diff
@@ -0,0 +1,372 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/impl/no_prelude/fail_impl_as_scope.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/impl/no_prelude/fail_impl_as_scope.carbon
+
+// --- fail_impl_as_file_scope.carbon
+library "[[@TEST_NAME]]";
+
+interface I {}
+
+// CHECK:STDERR: fail_impl_as_file_scope.carbon:[[@LINE+4]]:6: error: `impl as` can only be used in a class [ImplAsOutsideClass]
+// CHECK:STDERR: impl as I {}
+// CHECK:STDERR:      ^~
+// CHECK:STDERR:
+impl as I {}
+
+// --- fail_impl_as_function_scope.carbon
+library "[[@TEST_NAME]]";
+
+interface J {}
+
+fn F() {
+  // CHECK:STDERR: fail_impl_as_function_scope.carbon:[[@LINE+4]]:8: error: `impl as` can only be used in a class [ImplAsOutsideClass]
+  // CHECK:STDERR:   impl as J {}
+  // CHECK:STDERR:        ^~
+  // CHECK:STDERR:
+  impl as J {}
+}
+
+// --- fail_impl_as_self_interface.carbon
+library "[[@TEST_NAME]]";
+
+interface Z {
+  fn Zero();
+
+   // CHECK:STDERR: fail_impl_as_self_interface.carbon:[[@LINE+4]]:9: error: `impl as` can only be used in a class [ImplAsOutsideClass]
+   // CHECK:STDERR:    impl as Z {
+   // CHECK:STDERR:         ^~
+   // CHECK:STDERR:
+   impl as Z {
+    fn Zero() {}
+  }
+}
+
+class Point {
+   impl as Z {
+    fn Zero() {}
+  }
+}
+
+fn F() {
+  // Even if the `impl` is diagnosed above, we must not add the impl of the
+  // interface to itself in a way that allows it to be used during impl lookup,
+  // or we end up with infinite impl lookup recursion here.
+  ({} as Point).(Z.Zero)();
+}
+
+// --- fail_impl_as_other_interface.carbon
+
+library "[[@TEST_NAME]]";
+
+interface A {
+  fn B();
+}
+
+class X {
+  impl as A {
+    fn B() {}
+    // CHECK:STDERR: fail_impl_as_other_interface.carbon:[[@LINE+8]]:10: error: `impl as` can only be used in a class [ImplAsOutsideClass]
+    // CHECK:STDERR:     impl as B {}
+    // CHECK:STDERR:          ^~
+    // CHECK:STDERR:
+    // CHECK:STDERR: fail_impl_as_other_interface.carbon:[[@LINE+4]]:13: error: expression cannot be used as a value [UseOfNonExprAsValue]
+    // CHECK:STDERR:     impl as B {}
+    // CHECK:STDERR:             ^
+    // CHECK:STDERR:
+    impl as B {}
```

Comment:
What happens if this is `impl as C` instead of having a name collision with an interface member?

Reply:
Then we just get the one error - which is a cleaner test output - but also doesn't demonstrate it kept going but maybe that's okay here. I will make that change lmk which you prefer.

Reply:
Maybe split this into multiple tests, each focused on testing one thing? If you are concerned about showing that it continues to check, perhaps move the `fn B() {}` definition after this?

Reply:
I ensured/show that we continue checking by adding some simple semir after the invalid `impl as` in each test - a fn call to G().